Sources/LeafKit/LeafLexer.swift

```diff
@@ -1,147 +1,210 @@
                 }
                 // consume final quote
-                self.buffer.moveReaderIndex(forwardBy: 1)
+                pop()
                 return .stringLiteral(string)
             case .space:
                 // skip space
-                self.buffer.moveReaderIndex(forwardBy: 1)
-                return try self.next()
-            default:
-                guard let length = self.countMatching(check: { $0.isAllowedInVariable }) else {
-                    return nil
-                }
-                if length == 0 {
-                    let source = LeafSource.start(at: self.buffer).end(at: self.buffer)
-                    throw LeafError(.unexpectedToken, source: source)
-                }
-                guard let name = self.buffer.readString(length: length) else {
-                    return nil
-                }
+                pop()
+                return try self.nextToken()
+            case let x where x.isValidInParameter:
+                let read = readWhile { $0.isValidInParameter }
+                guard let name = read else { fatalError("switch case should disallow this") }
+                // this parameter is a tag
+                if peek() == .leftParenthesis { return .tag(name: name) }
+                
+                // check if expected parameter type
+                if let keyword = LeafToken.Keyword(rawValue: name) { return .keyword(keyword) }
+                else if let op = LeafToken.Operator(rawValue: name) { return .operator(op) }
+                else if let val = Int(name) { return .constant(.int(val)) }
+                else if let val = Double(name) { return .constant(.double(val)) }
+                
+                // unknown param type.. var
                 return .variable(name: name)
+            default:
+                fatalError("todo: unable to process throw error")
             }
+        case .body:
+            state = .normal
+            guard next == .colon else { return try nextToken() }
+            pop()
+            return .tagBodyIndicator
         }
     }
     
     // MARK: byte buffer methods
     
-    mutating func pop(if byte: UInt8) -> Bool {
-        if self.peek() == byte {
-            self.pop()
-            return true
-        } else {
-            return false
+    mutating func readWhile(_ check: (UInt8) -> Bool) -> String? {
+        guard let length = countMatching(check: check) else {
+            return nil
         }
+        return buffer.readString(length: length)
     }
     
-    func peek() -> UInt8? {
-        return self.buffer.getInteger(at: self.buffer.readerIndex)
+    mutating func readSliceWhile(_ check: (UInt8) -> Bool) -> ByteBuffer? {
+        guard let length = countMatching(check: check) else {
+            return nil
+        }
+        return buffer.readSlice(length: length)
+    }
+    
+    func peek(at idx: Int = 0) -> UInt8? {
```

Comment:
```suggestion
    func peek(aheadBy offset: Int = 0) -> UInt8? {
```

Reply:
was stuck what to name this, aheadBy is way better